ledger/participant-state-index/reference/src/main/scala/com/daml/ledger/participant/state/index/v1/impl/reference/ReferenceIndexService.scala

```diff
@@ -347,6 +347,27 @@ final case class ReferenceIndexService(
       }
     }
 
+  override def lookupKey(
+      submitter: Party,
+      key: GlobalKey): Future[Option[Value.AbsoluteContractId]] =
+    futureWithState { state =>
+      Future {
+
+        state.activeContracts.keys
+          .get(key)
+          .flatMap { cid =>
+            logger.debug(s"lookupKey: $submitter, $key: $cid")
+
+            // note that we need to check visibility for keys, too, otherwise we leak the existence of a non-divulged
+            // contract if we return `Some`.
+            state.activeContracts.lookupContract(cid).flatMap {
```

Comment:
DAML well-authorization rules state that fetchByKey fails and aborts the transaction if:
a) you don’t have sufficient authorization
b) you’re not a stakeholder of the contract you’re trying to fetch

In this context, it is not clear to me why wee need to call `canSeeContract`.
1.  Why do we need to call it explicitly here? I would expect the DAML interpreter to convert positive result to a negative one based on b) above
2. Why does submitter matter? Isn't it more about the actor of that specific fetch action?

Reply:
We need to authorize the lookup, otherwise the submitter could call `lookupByKey` to check for the existence of keys that it cannot access.

Reply:
See also the original implementation in Sandbox: https://github.com/digital-asset/daml/blob/master/ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/SandboxContractStore.scala#L33